File: built-in plugins dsp/mb-acid-bass.cpp
/**
 * MB Acid Bass
 * Category : instrument
 * Type     : synth
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Classic 303-style acid bass synthesizer with squelchy filter
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_ACID_BASS_H
#define MB_ACID_BASS_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbAcidBass : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-acid-bass";
    static constexpr const char* PLUGIN_NAME    = "MB Acid Bass";
    static constexpr const char* PLUGIN_TYPE    = "synth";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float cutoff = 400f;  // range [50, 5000]
    float resonance = 0.8f;  // range [0, 1]
    float env_mod = 0.7f;  // range [0, 1]
    float decay = 0.2f;  // range [0.01, 2]
    float accent = 0.5f;  // range [0, 1]
    float slide = 0.05f;  // range [0, 0.5]
    float distortion = 0.2f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbAcidBass() = default;
    ~MbAcidBass() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.cutoff = std::clamp(params.cutoff, 50f, 5000f);
        params.resonance = std::clamp(params.resonance, 0f, 1f);
        params.env_mod = std::clamp(params.env_mod, 0f, 1f);
        params.decay = std::clamp(params.decay, 0.01f, 2f);
        params.accent = std::clamp(params.accent, 0f, 1f);
        params.slide = std::clamp(params.slide, 0f, 0.5f);
        params.distortion = std::clamp(params.distortion, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Acid Bass
        return input;
    }
};

#endif // MB_ACID_BASS_H

File: built-in plugins dsp/mb-acid-bass.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbAcidBassPlugin: PluginDefinition = {
    id: 'mb-acid-bass',
    slug: 'mb-acid-bass',
    name: 'MB Acid Bass',
    category: 'instrument',
    type: 'synth',
    version: '1.0.0',
    description: 'Classic 303-style acid bass synthesizer with squelchy filter',
    author: 'Max Booster',
    oscillators: [
      { type: 'sawtooth', detune: 0, gain: 0.7 },
      { type: 'square', detune: 0, gain: 0.3 },
    ],
    envelope: { attack: 0.001, decay: 0.2, sustain: 0, release: 0.1 },
    parameters: [
      { id: 'waveform', name: 'Waveform', type: 'choice', defaultValue: 'sawtooth', choices: ['sawtooth', 'square'], automatable: false },
      { id: 'cutoff', name: 'Filter Cutoff', type: 'float', defaultValue: 400, minValue: 50, maxValue: 5000, unit: 'Hz', automatable: true },
      { id: 'resonance', name: 'Resonance', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true },
      { id: 'env_mod', name: 'Envelope Mod', type: 'float', defaultValue: 0.7, minValue: 0, maxValue: 1, automatable: true },
      { id: 'decay', name: 'Decay', type: 'float', defaultValue: 0.2, minValue: 0.01, maxValue: 2, unit: 's', automatable: true },
      { id: 'accent', name: 'Accent', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true },
      { id: 'slide', name: 'Slide Time', type: 'float', defaultValue: 0.05, minValue: 0, maxValue: 0.5, unit: 's', automatable: true },
      { id: 'distortion', name: 'Distortion', type: 'float', defaultValue: 0.2, minValue: 0, maxValue: 1, automatable: true },
      { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { waveform: 'sawtooth', cutoff: 400, resonance: 0.8, env_mod: 0.7, decay: 0.2, accent: 0.5, slide: 0.05, distortion: 0.2, volume: 0.8 },
  };

export default MbAcidBassPlugin;
